debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java

```diff
@@ -125,16 +126,25 @@ public void execute(ChangeEventSourceContext context) throws InterruptedExceptio
                 if (connectorConfig.isReadOnlyDatabaseConnection()) {
                     dataConnection.commit();
                 }
-                final MaxLsnResult maxLsnResult = dataConnection.getMaxLsnResult(connectorConfig.isSkipLowActivityLsnsEnabled());
+
+                Lsn lastProcessedLsn = lastProcessedPosition.getCommitLsn();
+                Lsn currentMaxLsn;
+
+                if (maxTransactionsPerIteration > 0) {
+                    currentMaxLsn = dataConnection.getNthTransactionLsn(lastProcessedLsn, maxTransactionsPerIteration);
+                }
+                else {
+                    currentMaxLsn = dataConnection.getMaxTransactionLsn();
+                }
 
                 // Shouldn't happen if the agent is running, but it is better to guard against such situation
-                if (!maxLsnResult.getMaxLsn().isAvailable() || !maxLsnResult.getMaxTransactionalLsn().isAvailable()) {
+                if (!currentMaxLsn.isAvailable()) {
                     LOGGER.warn("No maximum LSN recorded in the database; please ensure that the SQL Server Agent is running");
                     metronome.pause();
                     continue;
                 }
                 // There is no change in the database
-                if (maxLsnResult.getMaxTransactionalLsn().compareTo(lastProcessedPosition.getCommitLsn()) <= 0 && shouldIncreaseFromLsn) {
```

Comment:
Removing `&& shouldIncreaseFromLsn` makes the tests fail. Also it looks like adding `&& shouldIncreaseFromLsn` back is not the right fix as it might be fixing tests as a side effect. 

The real cause of the failures is the following. Suppose there was a transaction `LSN100` with 10 rows updated. The connector consumed only 5 of them and was interrupted. So the recorded possition is going to be `LSN100(5)`. When the connector is switched on again it needs to consume the remaining 5 rows of `LSN100`. If there were no new transactions in DB `currentMaxLsn` is going to be eq to `LSN100`, `lastProcessedPosition` is also `LSN100`. This `if` evaluates to true and skips the remaining 5 rows.

Reply:
But `shouldIncreaseFromLsn` is set to `true` only once the new max LSN is greater than the `lastProcessedPosition` (which will happen only when there _are_ new changes). Isn't it?

Reply:
I can't fully understand how the logic around `shouldIncreaseFromLsn` works. I see only two writes to this variable, both of which always write true (assuming `offsetContext.isSnapshotCompleted()` always returns `true` since the snapshot must be completed before the streaming phase). I don't see how `shouldIncreaseFromLsn` could be set to `false`. But while trying to fix the unit tests I found that bringing back `&& shouldIncreaseFromLsn` fixes them. This is why I say `is not the right fix`. But maybe I'm missing some detail here and this is indeed the right fix.

Reply:
As agreed offline, even if the issue above exists, it doesn't seem to be caused by this patch.

@ramanenka let's proceed with merging it as is and bring this question up when contributing upstream.